pylivetrader/backend/alpaca.py

```diff
@@ -495,48 +499,49 @@ def get_bars(self, assets, data_frequency, bar_count=500):
             symbols = [assets.symbol]
         else:
             symbols = [asset.symbol for asset in assets]
+        # the timezone of end_dt if UTC, when data_frequency=daily, end_dt is midnight in UTC time
+        if not end_dt:
+            end_dt = pd.to_datetime('now', utc=True).floor('min')
+            session_label = self._cal.minute_to_session_label(end_dt)
+            session_open = self._cal.session_open(session_label)
+            session_close = self._cal.session_close(session_label)
+            if end_dt < session_open:  # before market open
+                end_dt = self._cal.previous_minute(end_dt)
+            elif session_open <= end_dt < session_close:
+                pass
+            else:
+                end_dt = session_close
+            if is_daily:
+                end_dt = session_label
 
-        symbol_bars = self._symbol_bars(
-            symbols, 'day' if is_daily else 'minute', limit=bar_count)
-
+        all_minutes: pd.DatetimeIndex = self._cal.all_minutes
+        all_sessions: pd.DatetimeIndex = self._cal.all_sessions
         if is_daily:
-            intra_bars = {}
-            symbol_bars_minute = self._symbol_bars(
-                symbols, 'minute', limit=1000)
-            for symbol, df in symbol_bars_minute.items():
-                agged = df.resample('1D').agg(dict(
-                    open='first',
-                    high='max',
-                    low='min',
-                    close='last',
-                    volume='sum',
-                )).dropna()
-                intra_bars[symbol] = agged
-
-        dfs = []
-        for asset in assets if not assets_is_scalar else [assets]:
-            symbol = asset.symbol
-            df = symbol_bars.get(symbol)
-            if df is None:
-                dfs.append(pd.DataFrame(
-                    [], columns=[
-                        'open', 'high', 'low', 'close', 'volume']
-                ))
-                continue
-            if is_daily:
-                agged = intra_bars.get(symbol)
-                if agged is not None and len(
-                        agged.index) > 0 and agged.index[-1] not in df.index:
-                    if not (agged.index[-1] > df.index[-1]):
-                        log.warn(
-                            ('agged.index[-1] = {}, df.index[-1] = {} '
-                             'for {}').format(
-                                agged.index[-1], df.index[-1], symbol))
-                    df = df.append(agged.iloc[-1])
-            df.columns = pd.MultiIndex.from_product([[asset, ], df.columns])
-            dfs.append(df)
-
-        return pd.concat(dfs, axis=1)
+            e_idx = all_sessions.get_loc(end_dt)
```

Comment:
this is a HUGE block of code.
when reading code it should be like a story.
you should wrap it in a function. give a function name and document the function explaining exactly what you are doing.
the reader shouldn't try hard to understand